File: api/recognition_admin.php
<?php
// api/recognition_admin.php - 同好会运营：提交审核 / 兑换码管理
// 动作: submissions / review / claim_generate / claim_list
// 审核与兑换码生成均要求该同好会的相应角色权限。

header('Content-Type: application/json; charset=utf-8');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(204);
    exit;
}

require_once __DIR__ . '/../includes/auth.php';
require_once __DIR__ . '/../includes/audit.php';
require_once __DIR__ . '/../includes/display_club.php';
require_once __DIR__ . '/../includes/recognition/roles.php';
require_once __DIR__ . '/../includes/recognition/events.php';
require_once __DIR__ . '/../includes/recognition/pipeline.php';
require_once __DIR__ . '/../includes/recognition/credential.php';

function recogRespond(array $payload, int $code = 200): void {
    http_response_code($code);
    echo json_encode($payload, JSON_UNESCAPED_UNICODE);
    exit;
}

/**
 * 生成不含易混淆字符的兑换码
 */
function recogGenerateClaimCode(): string {
    $alphabet = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
    $code = '';
    for ($i = 0; $i < 10; $i++) {
        $code .= $alphabet[random_int(0, strlen($alphabet) - 1)];
    }
    return $code;
}

$action = $_GET['action'] ?? '';
$db = getDB();

switch ($action) {

    // ---- 待审核提交列表 ----
    case 'submissions': {
        $user = requireLogin();
        $clubId = (int)($_GET['club_id'] ?? 0);
        $country = displayClubCountry((string)($_GET['country'] ?? 'china')) ?? 'china';
        if (!recogCanReview($user, $clubId, $country) && $user['role'] !== 'super_admin') {
            recogRespond(['success' => false, 'message' => '权限不足'], 403);
        }
        $status = in_array($_GET['status'] ?? '', ['pending', 'approved', 'rejected'], true) ? $_GET['status'] : 'pending';

        $stmt = $db->prepare(
            "SELECT s.id, s.content, s.file_path, s.status, s.created_at,
                    v.id AS program_version_id, v.program_id,
                    p.title AS program_title,
                    u.username AS holder_username, u.nickname AS holder_nickname, u.id AS holder_user_id
             FROM recognition_submissions s
             JOIN recognition_program_versions v ON v.id = s.program_version_id
             JOIN recognition_programs p ON p.id = v.program_id
             JOIN users u ON u.id = s.user_id
             WHERE p.club_id = ? AND p.country = ? AND s.status = ?
             ORDER BY s.created_at ASC LIMIT 200"
        );
        $stmt->execute([$clubId, $country, $status]);
        recogRespond(['success' => true, 'submissions' => $stmt->fetchAll()]);
    }

    // ---- 审核通过 / 驳回（通过后触发规则评估与签发） ----
    case 'review': {
        if ($_SERVER['REQUEST_METHOD'] !== 'POST') recogRespond(['success' => false, 'message' => '仅支持 POST']);
        $user = requireLogin();
        $input = json_decode(file_get_contents('php://input'), true) ?: [];
        $submissionId = (int)($input['submission_id'] ?? 0);
        $decision = in_array($input['decision'] ?? '', ['approved', 'rejected'], true) ? $input['decision'] : '';
        $comment = trim((string)($input['comment'] ?? ''));
        if (!$decision) recogRespond(['success' => false, 'message' => '审核结论非法']);

        $stmt = $db->prepare(
            'SELECT s.*, v.id AS program_version_id, v.program_id
             FROM recognition_submissions s
             JOIN recognition_program_versions v ON v.id = s.program_version_id
             WHERE s.id = ?'
        );
        $stmt->execute([$submissionId]);
        $submission = $stmt->fetch();
        if (!$submission) recogRespond(['success' => false, 'message' => '提交不存在'], 404);
        if ($submission['status'] !== 'pending') recogRespond(['success' => false, 'message' => '该提交已审核']);

        $stmt = $db->prepare('SELECT * FROM recognition_programs WHERE id = ?');
        $stmt->execute([(int)$submission['program_id']]);
        $program = $stmt->fetch();
        if (!recogCanReview($user, (int)$program['club_id'], (string)$program['country']) && $user['role'] !== 'super_admin') {
            recogRespond(['success' => false, 'message' => '无权审核该提交'], 403);
        }

        $db->beginTransaction();
        try {
            $db->prepare('UPDATE recognition_submissions SET status = ?, reviewed_at = ? WHERE id = ?')
                ->execute([$decision, date('Y-m-d H:i:s'), $submissionId]);
            $db->prepare('INSERT INTO recognition_reviews (submission_id, reviewer_id, decision, comment) VALUES (?, ?, ?, ?)')
                ->execute([$submissionId, (int)$user['id'], $decision, $comment]);
            $db->commit();
        } catch (Throwable $e) {
            $db->rollBack();
            error_log('[recog] review failed: ' . $e->getMessage());
            recogRespond(['success' => false, 'message' => '审核失败'], 500);
        }

        $eventType = $decision === 'approved' ? 'submission.approved' : 'submission.rejected';
        recogRecordEvent([
            'type' => $eventType,
            'idempotency_key' => 'review_' . $submissionId,
            'club_id' => (int)$program['club_id'],
            'country' => (string)$program['country'],
            'user_id' => (int)$submission['user_id'],
            'program_id' => (int)$submission['program_id'],
            'program_version_id' => (int)$submission['program_version_id'],
            'data' => ['submission_id' => $submissionId, 'reviewer_id' => (int)$user['id']],
            'source_verified' => true,
        ]);
        logAction('recog_submission_' . $decision, 'recognition_submission', $submissionId, ['reviewer_id' => (int)$user['id']]);

        $award = ['passed' => false, 'issued' => false, 'credential' => null, 'duplicate' => false, 'error' => null];
        if ($decision === 'approved') {
            // 审核通过 → 规则评估（含提交通过与审核人数条件）
            $stmt = $db->prepare('SELECT COUNT(*) AS c FROM recognition_reviews WHERE submission_id = ? AND decision = ?');
            $stmt->execute([$submissionId, 'approved']);
            $reviewerCount = (int)$stmt->fetch()['c'];

            $award = recogEvaluateAndAward((int)$submission['program_version_id'], (int)$submission['user_id'], [
                'submission_status' => 'approved',
                'reviewer_count' => $reviewerCount,
                'source' => 'site',
            ]);
        }

        recogRespond([
            'success' => true,
            'decision' => $decision,
            'issued' => $award['issued'],
            'already_held' => $award['duplicate'],
            'credential' => $award['credential'],
            'message' => $decision === 'approved'
                ? ($award['error'] ?: ($award['passed'] ? '已通过并签发凭证' : '已通过，未满足签发条件'))
                : '已驳回',
        ]);
    }

    // ---- 生成兑换码（线下活动先参与、后领取） ----
    case 'claim_generate': {
        if ($_SERVER['REQUEST_METHOD'] !== 'POST') recogRespond(['success' => false, 'message' => '仅支持 POST']);
        $user = requireLogin();
        $input = json_decode(file_get_contents('php://input'), true) ?: [];
        $programId = (int)($input['program_id'] ?? 0);
        $count = min(500, max(1, (int)($input['count'] ?? 0)));
        $ttlHours = max(0, (int)($input['ttl_hours'] ?? 0));

        $stmt = $db->prepare('SELECT * FROM recognition_programs WHERE id = ?');
        $stmt->execute([$programId]);
        $program = $stmt->fetch();
        if (!$program) recogRespond(['success' => false, 'message' => '项目不存在'], 404);
        if (!recogCanIssue($user, (int)$program['club_id'], (string)$program['country']) && $user['role'] !== 'super_admin') {
            recogRespond(['success' => false, 'message' => '无权为该项目生成兑换码'], 403);
        }
        $version = recogPublishedVersion($db, $programId);
        if (!$version) recogRespond(['success' => false, 'message' => '该项目尚无已发布版本']);
        $badgeId = (int)($version['content']['rules']['award']['badge_id'] ?? 0);
        if ($badgeId <= 0) recogRespond(['success' => false, 'message' => '项目未配置奖励徽章']);

        $codes = [];
        $expiresAt = $ttlHours > 0 ? date('Y-m-d H:i:s', time() + $ttlHours * 3600) : null;
        $stmt = $db->prepare(
            'INSERT INTO recognition_claim_codes (code, program_id, program_version_id, badge_id, expires_at)
             VALUES (?, ?, ?, ?, ?)'
        );
        for ($i = 0; $i < $count; $i++) {
            // 碰撞重试
            for ($retry = 0; $retry < 5; $retry++) {
                $code = recogGenerateClaimCode();
                try {
                    $stmt->execute([$code, $programId, (int)$version['id'], $badgeId, $expiresAt]);
                    $codes[] = $code;
                    break;
                } catch (PDOException $e) {
                    if ($retry === 4) throw $e;
                }
            }
        }
        logAction('recog_claim_generated', 'recognition_program', $programId, ['count' => count($codes)]);
        recogRespond(['success' => true, 'codes' => $codes, 'expires_at' => $expiresAt]);
    }

    // ---- 批量导入活动参与名单（CSV 名单 → 直接签发，对齐架构文档 4.1 标准能力）----
    // 输入：usernames（数组）或 csv（用户名每行一个的文本，兼容逗号分隔）
    case 'import_participants': {
        if ($_SERVER['REQUEST_METHOD'] !== 'POST') recogRespond(['success' => false, 'message' => '仅支持 POST']);
        $user = requireLogin();
        $input = json_decode(file_get_contents('php://input'), true) ?: [];
        $programId = (int)($input['program_id'] ?? 0);

        $usernames = (array)($input['usernames'] ?? []);
        if (!$usernames && !empty($input['csv'])) {
            // 兼容 CSV 粘贴：支持换行与逗号两种分隔，去除可能的表头列分隔符之外的引号/空格/UTF-8 BOM
            $raw = preg_replace('/^\xEF\xBB\xBF/', '', (string)$input['csv']);
            $usernames = preg_split('/[\r\n,]+/', $raw);
        }
        $usernames = array_values(array_filter(array_map(function ($s) {
            return trim((string)$s, " \t\"'");
        }, $usernames)));

        if (!$usernames) recogRespond(['success' => false, 'message' => '请提供参与名单（用户名列表或 CSV 文本）']);
        if (count($usernames) > 2000) recogRespond(['success' => false, 'message' => '单次导入最多 2000 人']);

        $stmt = $db->prepare('SELECT * FROM recognition_programs WHERE id = ?');
        $stmt->execute([$programId]);
        $program = $stmt->fetch();
        if (!$program) recogRespond(['success' => false, 'message' => '项目不存在'], 404);
        if (!in_array($program['type'], ['activity', 'award'], true)) {
            recogRespond(['success' => false, 'message' => '名单导入仅支持 activity / award 类型项目']);
        }
        if (!recogCanIssue($user, (int)$program['club_id'], (string)$program['country']) && $user['role'] !== 'super_admin') {
            recogRespond(['success' => false, 'message' => '无权为该项目导入名单'], 403);
        }
        $version = recogPublishedVersion($db, $programId);
        if (!$version) recogRespond(['success' => false, 'message' => '该项目尚无已发布版本']);
        $badgeId = (int)($version['content']['rules']['award']['badge_id'] ?? 0);
        if ($badgeId <= 0) recogRespond(['success' => false, 'message' => '项目未配置奖励徽章']);

        $imported = 0; $skipped = [];
        foreach ($usernames as $username) {
            $stmt = $db->prepare("SELECT id FROM users WHERE username = ? AND status = 'active'");
            $stmt->execute([$username]);
            $target = $stmt->fetch();
            if (!$target) { $skipped[] = $username . '（账号不存在）'; continue; }

            // activity.attended 事实事件（幂等键绑定导入名单条目）
            recogRecordEvent([
                'type' => 'activity.attended',
                'idempotency_key' => 'import_' . $version['id'] . '_' . $target['id'],
                'club_id' => (int)$program['club_id'],
                'country' => (string)$program['country'],
                'user_id' => (int)$target['id'],
                'program_id' => $programId,
                'program_version_id' => (int)$version['id'],
                'badge_id' => $badgeId,
                'data' => ['imported_by' => (int)$user['id']],
                'source_verified' => true,
            ]);

            $issue = issueCredential($db, [
                'holder_user_id' => (int)$target['id'],
                'badge_id' => $badgeId,
                'issuer_club_id' => (int)$program['club_id'],
                'issuer_country' => (string)$program['country'],
                'program_id' => $programId,
                'program_version_id' => (int)$version['id'],
                'credential_type' => 'participation',
                'verification_level' => 'batch_import',
                'condition_summary' => ['活动参与名单批量导入'],
                'source_type' => 'batch_import',
                'actor_user_id' => (int)$user['id'],
            ]);
            if ($issue['ok'] && !$issue['duplicate']) $imported++;
            if ($issue['duplicate']) $skipped[] = $username . '（已持有）';
        }

        logAction('recog_participants_imported', 'recognition_program', $programId, [
            'imported' => $imported, 'skipped' => count($skipped),
        ]);
        recogRespond(['success' => true, 'imported' => $imported, 'skipped' => array_slice($skipped, 0, 50), 'skipped_total' => count($skipped)]);
    }

    // ---- 兑换码核销情况 ----
    case 'claim_list': {
        $user = requireLogin();
        $programId = (int)($_GET['program_id'] ?? 0);
        $stmt = $db->prepare('SELECT * FROM recognition_programs WHERE id = ?');
        $stmt->execute([$programId]);
        $program = $stmt->fetch();
        if (!$program) recogRespond(['success' => false, 'message' => '项目不存在'], 404);
        if (!recogCanIssue($user, (int)$program['club_id'], (string)$program['country']) && $user['role'] !== 'super_admin') {
            recogRespond(['success' => false, 'message' => '权限不足'], 403);
        }
        $stmt = $db->prepare(
            'SELECT code, redeemed_by, redeemed_at, expires_at FROM recognition_claim_codes
             WHERE program_id = ? ORDER BY id DESC LIMIT 1000'
        );
        $stmt->execute([$programId]);
        $rows = $stmt->fetchAll();
        $redeemed = count(array_filter($rows, fn($r) => $r['redeemed_by'] !== null));
        recogRespond(['success' => true, 'total' => count($rows), 'redeemed' => $redeemed, 'codes' => $rows]);
    }

    default:
        recogRespond(['success' => false, 'message' => '未知操作'], 400);
}

File: backend/internal/httpapi/recognition_admin.go
package httpapi

import (
	"crypto/rand"
	"database/sql"
	"net/http"
	"strings"
	"time"
)

const recognitionClaimAlphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

func (s *Server) recognitionAdmin(w http.ResponseWriter, r *http.Request) {
	publicAPIHeaders(w)
	if r.Method == http.MethodOptions {
		w.WriteHeader(http.StatusNoContent)
		return
	}
	switch strings.ToLower(strings.TrimSpace(r.URL.Query().Get("action"))) {
	case "submissions":
		s.recognitionAdminSubmissions(w, r)
	case "review":
		s.recognitionAdminReview(w, r)
	case "claim_generate":
		s.recognitionAdminClaimGenerate(w, r)
	case "import_participants":
		s.recognitionAdminImportParticipants(w, r)
	case "claim_list":
		s.recognitionAdminClaimList(w, r)
	default:
		writeJSONStatus(w, http.StatusBadRequest, map[string]any{"success": false, "message": "未知操作"})
	}
}

func (s *Server) recognitionAdminSubmissions(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		methodNotAllowed(w, http.MethodGet)
		return
	}
	user, ok := s.clubCodeUserResponse(w, r)
	if !ok {
		return
	}
	clubID := parsePositiveInt(r.URL.Query().Get("club_id"))
	country := clubCodeCountry(r.URL.Query().Get("country"))
	if !s.recognitionCanReview(r.Context(), user, clubID, country) {
		writeJSONStatus(w, http.StatusForbidden, map[string]any{"success": false, "message": "权限不足"})
		return
	}
	status := stringValue(r.URL.Query().Get("status"))
	if !containsString([]string{"pending", "approved", "rejected"}, status) {
		status = "pending"
	}
	rows, err := s.db.QueryContext(r.Context(), `SELECT s.id,s.content,s.file_path,s.status,s.created_at,v.id,v.program_id,p.title,u.username,COALESCE(u.nickname,''),u.id
        FROM recognition_submissions s JOIN recognition_program_versions v ON v.id=s.program_version_id JOIN recognition_programs p ON p.id=v.program_id
        JOIN users u ON u.id=s.user_id WHERE p.club_id=? AND p.country=? AND s.status=? ORDER BY s.created_at ASC LIMIT 200`, clubID, country, status)
	if err != nil {
		writeJSONStatus(w, http.StatusServiceUnavailable, map[string]any{"success": false, "message": "读取提交失败"})
		return
	}
	defer rows.Close()
	result := []map[string]any{}
	for rows.Next() {
		var id, versionID, programID, holderID int64
		var content, filePath, rowStatus, created, title, username, nickname string
		if err := rows.Scan(&id, &content, &filePath, &rowStatus, &created, &versionID, &programID, &title, &username, &nickname, &holderID); err != nil {
			continue
		}
		result = append(result, map[string]any{"id": id, "content": content, "file_path": filePath, "status": rowStatus, "created_at": created, "program_version_id": versionID, "program_id": programID, "program_title": title, "holder_username": username, "holder_nickname": nickname, "holder_user_id": holderID})
	}
	writeJSON(w, map[string]any{"success": true, "submissions": result})
}

func (s *Server) recognitionAdminReview(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSON(w, map[string]any{"success": false, "message": "仅支持 POST"})
		return
	}
	user, ok := s.clubCodeUserResponse(w, r)
	if !ok {
		return
	}
	var input map[string]any
	if err := decodeJSON(r, &input, 1<<20); err != nil || input == nil {
		writeJSON(w, map[string]any{"success": false, "message": "请求数据格式错误"})
		return
	}
	submissionID := integerValue(input["submission_id"])
	decision := stringValue(input["decision"])
	if !containsString([]string{"approved", "rejected"}, decision) {
		writeJSON(w, map[string]any{"success": false, "message": "审核结论非法"})
		return
	}
	var programID, versionID, holderID, clubID int64
	var country, currentStatus string
	if err := s.db.QueryRowContext(r.Context(), `SELECT s.user_id,s.status,v.id,v.program_id,p.club_id,p.country FROM recognition_submissions s JOIN recognition_program_versions v ON v.id=s.program_version_id JOIN recognition_programs p ON p.id=v.program_id WHERE s.id=?`, submissionID).Scan(&holderID, &currentStatus, &versionID, &programID, &clubID, &country); err != nil {
		writeJSONStatus(w, http.StatusNotFound, map[string]any{"success": false, "message": "提交不存在"})
		return
	}
	if currentStatus != "pending" {
		writeJSON(w, map[string]any{"success": false, "message": "该提交已审核"})
		return
	}
	if !s.recognitionCanReview(r.Context(), user, clubID, country) {
		writeJSONStatus(w, http.StatusForbidden, map[string]any{"success": false, "message": "无权审核该提交"})
		return
	}
	comment := truncateRecognition(strings.TrimSpace(stringValue(input["comment"])), 5000)
	tx, err := s.db.BeginTx(r.Context(), nil)
	if err != nil {
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "审核失败"})
		return
	}
	if _, err = tx.ExecContext(r.Context(), "UPDATE recognition_submissions SET status=?,reviewed_at=CURRENT_TIMESTAMP WHERE id=? AND status='pending'", decision, submissionID); err != nil {
		_ = tx.Rollback()
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "审核失败"})
		return
	}
	if _, err = tx.ExecContext(r.Context(), "INSERT INTO recognition_reviews(submission_id,reviewer_id,decision,comment) VALUES(?,?,?,?)", submissionID, user.ID, decision, comment); err != nil {
		_ = tx.Rollback()
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "审核失败"})
		return
	}
	eventType := "submission.rejected"
	if decision == "approved" {
		eventType = "submission.approved"
	}
	if _, err = recordRecognitionEventExec(r.Context(), tx, map[string]any{"type": eventType, "idempotency_key": "review_" + strconvInt(submissionID), "club_id": clubID, "country": country, "user_id": holderID, "program_id": programID, "program_version_id": versionID, "data": map[string]any{"submission_id": submissionID, "reviewer_id": user.ID}, "source_verified": true}); err != nil {
		_ = tx.Rollback()
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "审核失败"})
		return
	}
	if err = tx.Commit(); err != nil {
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "审核失败"})
		return
	}
	award := map[string]any{"passed": false, "issued": false, "duplicate": false, "credential": nil, "error": nil}
	if decision == "approved" {
		var reviewerCount int64
		_ = s.db.QueryRowContext(r.Context(), "SELECT COUNT(*) FROM recognition_reviews WHERE submission_id=? AND decision='approved'", submissionID).Scan(&reviewerCount)
		award = s.recognitionEvaluateAndAward(r.Context(), versionID, holderID, map[string]any{"submission_status": "approved", "reviewer_count": reviewerCount, "source": "site"})
	}
	message := "已驳回"
	if decision == "approved" {
		message = firstNonEmpty(stringValue(award["error"]), map[bool]string{true: "已通过并签发凭证", false: "已通过，未满足签发条件"}[boolValue(award["passed"])])
	}
	writeJSON(w, map[string]any{"success": true, "decision": decision, "issued": boolValue(award["issued"]), "already_held": boolValue(award["duplicate"]), "credential": award["credential"], "message": message})
}

func (s *Server) recognitionAdminClaimGenerate(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSON(w, map[string]any{"success": false, "message": "仅支持 POST"})
		return
	}
	user, ok := s.clubCodeUserResponse(w, r)
	if !ok {
		return
	}
	var input map[string]any
	if err := decodeJSON(r, &input, 1<<20); err != nil || input == nil {
		writeJSON(w, map[string]any{"success": false, "message": "请求数据格式错误"})
		return
	}
	programID := integerValue(input["program_id"])
	count := integerValue(input["count"])
	if count < 1 {
		count = 1
	}
	if count > 500 {
		count = 500
	}
	ttlHours := maxInt64(integerValue(input["ttl_hours"]), 0)
	program, err := s.recognitionProgram(r.Context(), programID)
	if err != nil || program == nil {
		writeJSONStatus(w, http.StatusNotFound, map[string]any{"success": false, "message": "项目不存在"})
		return
	}
	if !s.recognitionCanIssue(r.Context(), user, program.ClubID, program.Country) {
		writeJSONStatus(w, http.StatusForbidden, map[string]any{"success": false, "message": "无权为该项目生成兑换码"})
		return
	}
	version, err := s.recognitionVersion(r.Context(), programID, false)
	if err != nil || version == nil {
		writeJSON(w, map[string]any{"success": false, "message": "该项目尚无已发布版本"})
		return
	}
	content, _ := version["content"].(map[string]any)
	rules, _ := content["rules"].(map[string]any)
	award, _ := rules["award"].(map[string]any)
	badgeID := integerValue(award["badge_id"])
	if badgeID <= 0 {
		writeJSON(w, map[string]any{"success": false, "message": "项目未配置奖励徽章"})
		return
	}
	var expires any
	var expiresText any
	if ttlHours > 0 {
		expiry := time.Now().Add(time.Duration(ttlHours) * time.Hour).Format("2006-01-02 15:04:05")
		expires, expiresText = expiry, expiry
	}
	tx, err := s.db.BeginTx(r.Context(), nil)
	if err != nil {
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "兑换码生成失败"})
		return
	}
	codes := []string{}
	for index := int64(0); index < count; index++ {
		inserted := false
		for retry := 0; retry < 5 && !inserted; retry++ {
			code, codeErr := newRecognitionClaimCode()
			if codeErr != nil {
				_ = tx.Rollback()
				writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "兑换码生成失败"})
				return
			}
			if _, err = tx.ExecContext(r.Context(), "INSERT INTO recognition_claim_codes(code,program_id,program_version_id,badge_id,expires_at) VALUES(?,?,?,?,?)", code, programID, integerValue(version["id"]), badgeID, expires); err == nil {
				codes = append(codes, code)
				inserted = true
			} else if !recognitionUniqueError(err) {
				_ = tx.Rollback()
				writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "兑换码生成失败"})
				return
			}
		}
		if !inserted {
			_ = tx.Rollback()
			writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "兑换码生成失败"})
			return
		}
	}
	if err = tx.Commit(); err != nil {
		writeJSONStatus(w, http.StatusInternalServerError, map[string]any{"success": false, "message": "兑换码生成失败"})
		return
	}
	writeJSON(w, map[string]any{"success": true, "codes": codes, "expires_at": expiresText})
}

func (s *Server) recognitionAdminImportParticipants(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeJSON(w, map[string]any{"success": false, "message": "仅支持 POST"})
		return
	}
	user, ok := s.clubCodeUserResponse(w, r)
	if !ok {
		return
	}
	var input map[string]any
	if err := decodeJSON(r, &input, 2<<20); err != nil || input == nil {
		writeJSON(w, map[string]any{"success": false, "message": "请求数据格式错误"})
		return
	}
	programID := integerValue(input["program_id"])
	usernames := stringSlice(input["usernames"], 2000)
	if len(usernames) == 0 {
		raw := strings.TrimPrefix(stringValue(input["csv"]), "\ufeff")
		for _, token := range strings.FieldsFunc(raw, func(r rune) bool { return r == '\n' || r == '\r' || r == ',' }) {
			if token = strings.Trim(token, " \t\"'"); token != "" {
				usernames = append(usernames, token)
			}
		}
	}
	if len(usernames) == 0 {
		writeJSON(w, map[string]any{"success": false, "message": "请提供参与名单（用户名列表或 CSV 文本）"})
		return
	}
	if len(usernames) > 2000 {
		writeJSON(w, map[string]any{"success": false, "message": "单次导入最多 2000 人"})
		return
	}
	program, err := s.recognitionProgram(r.Context(), programID)
	if err != nil || program == nil {
		writeJSONStatus(w, http.StatusNotFound, map[string]any{"success": false, "message": "项目不存在"})
		return
	}
	if program.Type != "activity" && program.Type != "award" {
		writeJSON(w, map[string]any{"success": false, "message": "名单导入仅支持 activity / award 类型项目"})
		return
	}
	if !s.recognitionCanIssue(r.Context(), user, program.ClubID, program.Country) {
		writeJSONStatus(w, http.StatusForbidden, map[string]any{"success": false, "message": "无权为该项目导入名单"})
		return
	}
	version, err := s.recognitionVersion(r.Context(), programID, false)
	if err != nil || version == nil {
		writeJSON(w, map[string]any{"success": false, "message": "该项目尚无已发布版本"})
		return
	}
	content, _ := version["content"].(map[string]any)
	rules, _ := content["rules"].(map[string]any)
	award, _ := rules["award"].(map[string]any)
	badgeID := integerValue(award["badge_id"])
	if badgeID <= 0 {
		writeJSON(w, map[string]any{"success": false, "message": "项目未配置奖励徽章"})
		return
	}
	imported, skipped := 0, []string{}
	for _, username := range usernames {
		target, targetErr := s.findUserByLogin(r.Context(), username)
		if targetErr != nil || target == nil {
			skipped = append(skipped, username+"（账号不存在）")
			continue
		}
		_, _ = recordRecognitionEventExec(r.Context(), s.db, map[string]any{"type": "activity.attended", "idempotency_key": "import_" + strconvInt(integerValue(version["id"])) + "_" + strconvInt(target.ID), "club_id": program.ClubID, "country": program.Country, "user_id": target.ID, "program_id": programID, "program_version_id": integerValue(version["id"]), "badge_id": badgeID, "data": map[string]any{"imported_by": user.ID}, "source_verified": true})
		_, duplicate, issueErr := s.issueRecognitionCredential(r.Context(), recognitionIssueArgs{HolderUserID: target.ID, BadgeID: badgeID, IssuerClubID: program.ClubID, IssuerCountry: program.Country, ProgramID: programID, ProgramVersionID: integerValue(version["id"]), CredentialType: "participation", VerificationLevel: "batch_import", Conditions: []any{"活动参与名单批量导入"}, SourceType: "batch_import", ActorUserID: user.ID})
		if issueErr != nil {
			skipped = append(skipped, username)
		} else if duplicate {
			skipped = append(skipped, username+"（已持有）")
		} else {
			imported++
		}
	}
	shown := skipped
	if len(shown) > 50 {
		shown = shown[:50]
	}
	writeJSON(w, map[string]any{"success": true, "imported": imported, "skipped": shown, "skipped_total": len(skipped)})
}

func (s *Server) recognitionAdminClaimList(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		methodNotAllowed(w, http.MethodGet)
		return
	}
	user, ok := s.clubCodeUserResponse(w, r)
	if !ok {
		return
	}
	programID := parsePositiveInt(r.URL.Query().Get("program_id"))
	program, err := s.recognitionProgram(r.Context(), programID)
	if err != nil || program == nil {
		writeJSONStatus(w, http.StatusNotFound, map[string]any{"success": false, "message": "项目不存在"})
		return
	}
	if !s.recognitionCanIssue(r.Context(), user, program.ClubID, program.Country) {
		writeJSONStatus(w, http.StatusForbidden, map[string]any{"success": false, "message": "权限不足"})
		return
	}
	rows, err := s.db.QueryContext(r.Context(), "SELECT code,redeemed_by,redeemed_at,expires_at FROM recognition_claim_codes WHERE program_id=? ORDER BY id DESC LIMIT 1000", programID)
	if err != nil {
		writeJSONStatus(w, http.StatusServiceUnavailable, map[string]any{"success": false, "message": "读取兑换码失败"})
		return
	}
	defer rows.Close()
	codes := []map[string]any{}
	redeemed := 0
	for rows.Next() {
		var code string
		var redeemedBy sql.NullInt64
		var redeemedAt, expiresAt sql.NullString
		if rows.Scan(&code, &redeemedBy, &redeemedAt, &expiresAt) != nil {
			continue
		}
		if redeemedBy.Valid {
			redeemed++
		}
		codes = append(codes, map[string]any{"code": code, "redeemed_by": nullableInt(redeemedBy), "redeemed_at": recognitionNull(redeemedAt), "expires_at": recognitionNull(expiresAt)})
	}
	writeJSON(w, map[string]any{"success": true, "total": len(codes), "redeemed": redeemed, "codes": codes})
}

func newRecognitionClaimCode() (string, error) {
	result := make([]byte, 10)
	raw := make([]byte, 10)
	if _, err := rand.Read(raw); err != nil {
		return "", err
	}
	for index, value := range raw {
		result[index] = recognitionClaimAlphabet[int(value)%len(recognitionClaimAlphabet)]
	}
	return string(result), nil
}
